File: sportsbook-app/betting-service/src/main/java/com/sportsbook/bettingservice/kafka/BetEventMapper.java
package com.sportsbook.bettingservice.kafka;

import com.sportsbook.bettingservice.domain.Bet;
import com.sportsbook.events.BetPlaced;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface BetEventMapper {

    // existing: local event -> Avro (used by BetEventProducer)
    @Mapping(target = "eventId",   expression = "java(java.util.UUID.randomUUID().toString())")
    @Mapping(target = "betId",     source = "id")
    @Mapping(target = "playerId",  source = "playerId")
    @Mapping(target = "userId",    source = "userId")
    @Mapping(target = "marketId",  source = "matchId")     // matchId -> marketId
    @Mapping(target = "outcomeId", source = "selection")   // selection -> outcomeId
    @Mapping(target = "stake",     source = "stake")
    @Mapping(target = "odds",      source = "odds")
    @Mapping(target = "timestamp", expression = "java(System.currentTimeMillis())")
    BetPlaced toAvro(BetPlacedEvent source);

    // NEW: domain Bet -> local BetPlacedEvent (used by service)
    @Mapping(target = "id",        source = "id")
    @Mapping(target = "playerId",  source = "playerId")
    @Mapping(target = "userId",    source = "userId")
    @Mapping(target = "matchId",   source = "marketId")    // marketId -> matchId
    @Mapping(target = "selection", source = "outcomeId")   // outcomeId -> selection
    @Mapping(target = "stake",     source = "stake")
    @Mapping(target = "odds",      source = "odds")
    BetPlacedEvent toLocalEvent(Bet bet);
}

File: sportsbook-app/betting-service/src/main/kotlin/com/sportsbook/bettingservice/kafka/BetEventMapper.kt
package com.sportsbook.bettingservice.kafka

import com.sportsbook.bettingservice.domain.Bet
import com.sportsbook.events.BetPlaced
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class BetEventMapper {

    fun toAvro(source: BetPlacedEvent): BetPlaced =
        BetPlaced.newBuilder()
            .setEventId(UUID.randomUUID().toString())
            .setBetId(source.id)
            .setPlayerId(source.playerId)
            .setUserId(source.userId)
            .setMarketId(source.matchId)
            .setOutcomeId(source.selection)
            .setStake(source.stake)
            .setOdds(source.odds)
            .setTimestamp(System.currentTimeMillis())
            .build()

    fun toLocalEvent(bet: Bet): BetPlacedEvent =
        BetPlacedEvent(
            id = bet.id,
            playerId = bet.playerId,
            userId = bet.userId,
            matchId = bet.marketId,
            selection = bet.outcomeId,
            stake = bet.stake,
            odds = bet.odds,
        )
}
